docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,187 @@
+such as LDAP.
+
+X.509 certificates Vs OpenID Connect
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here, each user owns a signed certificate that is validated by the Kubernetes
+API server.
+
+This approach is not user-friendly that is each certificate has to be manually
+signed. Providing certificates for accessing the MetalK8s UI needs
+much more efforts since these certificates are browser incompatible.
+Using certificates is tedious since the certificate revocation process is also
+cumbersome.
+
+Keycloak Vs Dex
+~~~~~~~~~~~~~~~
+
+Both systems use OpenID Connect(OIDC) to authenticate a user using a
+standard OAuth2 flow.
+
+They both offer the ability to have short lived sessions so that user access
+can be rotated with minimum efforts.
+
+Finally, they both provide a means for identity management to be handled by an
+external service such as LDAP, Active Directory, SAML and others.
+
+Why not `Keycloak <https://www.keycoak.org/>`_?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Keycloak while offering similar features as Dex and even much more was rejected
+for the following reasons:
+
+- Keycloak is complex to operate(requires its own standalone database) and
+  manage(frequent database backups are required).
+
+- Currently, no use case exist for implementing a sophisticated Identity
+  Provider like Keycloak when the minimal Identity Provider from Dex is
+  sufficient.
+
+Note that, Keycloak is considered a future fallback Identity Provider if
+the need ever arises from a customer standpoint.
+
+Unexploited choices
+~~~~~~~~~~~~~~~~~~~
+
+* `Guard <https://github.com/appscode/guard/>`_
+
+A Kubernetes webhook authentication server by AppsCode, allowing you to
+log into your Kubernetes cluster by using various identity providers such as
+LDAP.
+
+* `ORY Hydra <https://github.com/ory/hydra/>`_
+
+It's an OpenID Connect provider, optimized for low resource consumption.
+ORY Hydra is not an identity provider but it is able to connect to existing
+identity providers.
+
+Implementation Details
+----------------------
+
+Implementation steps(High level)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Iteration 1
+~~~~~~~~~~~
+
+- Using Salt, generate self-signed certificates needed for Dex deployment
+- Deploy Dex in MetalK8s from the official **Dex Charts** while making use of
+  the generated certificates above
+- Configure Kubernetes API server flags to use Dex
+- Expose Dex on the control plane using Ingress
+- Implement MetalK8s UI integration with Dex
+- Theme the Dex GUI to match MetalK8s UI specs
+
+Iteration 2
+~~~~~~~~~~~
+
+- Provide integration with an external Identity Provider focusing on
+  LDAP and Active Directory
+- Provide documentation on how to integrate with these external Identity
+  Providers
+
+Iteration 3
+~~~~~~~~~~~
+
+- Provide a CLI for user management and provisioning
+
+Iteration 4
+~~~~~~~~~~~
+
+- Provide UI integration that performs similar CLI operations for user
+  management and provisioning.
```

Comment:
Same as above, some details to help understand the complexity of the task. The first iteration above is good (generate certs, use Helm charts, configure API server...)